doc/reference/storage_zfs.md

```diff
@@ -5,70 +5,111 @@ discourse: 1333
+- A *ZFS snapshot* captures a specific state of either a ZFS filesystem or a ZFS volume.
+  ZFS snapshots are read-only.
+- A *ZFS clone* is a writable copy of a ZFS snapshot.
+
+## `zfs` driver in LXD
+
+The `zfs` driver in LXD uses ZFS filesystems for images, and ZFS snapshots and clones to create instances and snapshots.
+By default, LXD enables compression when creating a ZFS pool.
+
+LXD assumes that it has full control over the ZFS pool and dataset.
+Therefore, you should never maintain any datasets or file system entities that are not owned by LXD in a LXD ZFS pool or dataset, because LXD might delete them.
+
+Due to the way copy-on-write works in ZFS, parent ZFS filesystems can't be removed until all children are gone.
+As a result, LXD automatically renames any objects that are removed but still referenced.
+Such objects are kept at a random `deleted/` path until all references are gone and the object can safely be removed.
+
+LXD automatically enables trimming support on all newly created pools on ZFS 0.8 or later.
+This increases the lifetime of SSDs by allowing better block re-use by the controller, and it also allows to free space on the root filesystem when using a loop-backed ZFS pool.
+If you are running a ZFS version earlier than 0.8 and want to enable trimming, upgrade to at least version 0.8.
+Then use the following commands to make sure that trimming is automatically enabled for the ZFS pool in the future and trim all currently unused space:
+
+    zpool upgrade ZPOOL-NAME
+    zpool set autotrim=on ZPOOL-NAME
+    zpool trim ZPOOL-NAME
+
+### Limitations
+
+The `zfs` driver has the following limitations:
+
+Delegating part of a pool
+: ZFS doesn't support delegating part of a pool to a container user.
+  Upstream is actively working on providing this functionality.
+
+Restoring from older snapshots
+: ZFS doesn't support restoring from snapshots other than the latest one.
+  You can, however, create new instances from older snapshots.
+  This method makes it possible to confirm whether a specific snapshot contains what you need.
+  After determining the correct snapshot, you can {ref}`remove the newer snapshots <storage-edit-snapshots>` so that the snapshot you need is the latest one and you can restore it.
+
+  Alternatively, you can configure LXD to automatically discard the newer snapshots during restore.
+  To do so, set the {ref}`zfs.remove_snapshots <storage-zfs-vol-config>` configuration for the volume (or the corresponding `volume.zfs.remove_snapshots` configuration on the storage pool for all volumes in the pool).
+
+  Note, however, that instance copies use ZFS snapshots too.
+  Therefore, you cannot restore an instance to a snapshot taken before the last copy without having to also delete all its descendants.
+  If this is not an option, you can copy the wanted snapshot into a new instance and then delete the old instance.
+  You will, however, lose any other snapshots the instance might have had.
```

Comment:
Does that mean that `zfs.clone_copy` must be set to "false" or "rebase" so that the copying workaround will work?
Or that this limitation only applies if `zfs.clone_copy` is set to "false" or "rebase"?

Reply:
This was really supposed to be referencing the line above that says "Note, however, that instance copies use ZFS snapshots too.", this is true if `zfs.clone_copy` is set to the default `true`. However if set to `false` or `rebase` then instance copies are done by duplicating the volume (either fully in the case of `false` or in the case of `rebase` then using the source instance's original image snapshot as the basis of the copy).

In those circumstances, instance copies are not snapshots of the source instance, and thus they do not contribute toward any restriction with regard to restoring snapshots of the source instance.

Reply:
Thanks! Fixed.